src/compute/src/compute_state.rs

```diff
@@ -416,16 +423,22 @@ impl<'a, A: Allocate + 'static> ActiveComputeState<'a, A> {
             }
         }
 
-        // If this is a non-transient dataflow, suspend it until we have capacity to start its
-        // hydration. See `process_sequential_hydration`.
+        // Schedule transient dataflows immediately, under the assumption that they are created for
+        // interactive user queries. Suspend non-transient dataflows, to enforce the configured
+        // `hydration_concurrency`.
         let (start_signal, suspension_token) = StartSignal::new();
         if dataflow.is_transient() {
             drop(suspension_token);
```

Comment:
While I understand the check for transient dataflows, it feels odd to me to have it in this place. I think we're dealing with two different concepts, that from a high level might be mixed. There's controller admission control that influences when `Schedule` commands are sent, and there's replica-local admission control where the replica maintains a bounded queue of what's currently hydrating. The transient dataflow check allows the replica to bypass the local admission control, but it also bypasses the global admission control, which is something the controller might have thought about.

Could we adjust this to include a field on the schedule command that indicates a deadline when to schedule the dataflow at the latest? For simplicity reasons, a boolean would suffice at the moment, but a more flexible deadline would allow the controller to maintain more control over the order in which dataflows would be hydrated.

Reply:
I mostly agree with what you are saying. You are right about the check here being a smell. I let it here mostly because it was the easiest way to get the PR into a working state that can be merged.

I'm working on a follow up (https://github.com/MaterializeInc/materialize/pull/26131) that moves the entire sequential hydration logic into a `SequentialHydration` client, and thereby also resolves the smell. Are you okay with merging this PR in its current state and then resolving the concern in the follow-up PR? We can merge both but I'd prefer not to, to keep the diffs small.

> a more flexible deadline would allow the controller to maintain more control over the order in which dataflows would be hydrated

It's not clear to me how the controller would set such a deadline. Would it try to figure out how long hydration for a given workload should reasonably take somehow? Or would it just enforce an upper bound after which we disable sequential hydration? Also, what about replicas that get spawned after the deadline?